doc/pins/pins_hld.md

```diff
@@ -0,0 +1,194 @@
+
+### Application Level Responses
+PINS introduces the concept of application level responses. SONiC currently supports synchronous communication between SWSS and Syncd containers. PINS extends synchronized communication to applications.
+
+Most SDN controllers require an acknowledgement of success or failure for each programming request. Subsequent programming from the controller depends on the response. SDN controllers are intelligent and can adapt to failures quickly. If one of the switches encounters a failure, the controller can utilize the other switches to achieve its goals. 
+
+The controller requires state information from the switches in order to quickly identify any failures. This is accomplished by the addition of application level responses and state. More information regarding this is provided in the supplementary documentation [Response path and Application DB](#broken-todo).
+
+### Packet IO
+P4RT clients can program ACLs to punt packets received in the ingress pipeline. These packets will be trapped and punted to the P4RT application running in the switch and will be forwarded to the client over grpc. 
+
+SONiC relied on basic packet I/O support that uses netdev ports to send and receive packets. P4Runtime applications require additional packet attributes like the target egress port which are not available via netdev. Similarly directed Tx from application level is not currently supported in the SONiC framework. This is required to support scenarios where the remote SDN controller needs to send a packet via one of the front panel ports of the switch.
+
+To support the above requirements, a model similar to sFlow is used. The receive solution creates a netlink type host interface. It programs a user defined trap for packets that are interesting to the controller and maps them to the netlink host interface. A generic ASIC independent model is defined for passing parameters such as “target egress port” to the application container. A detailed description of the receive path is present in the supplementary document on [Packet IO](#broken-todo).
+
+### Repositories
+This design adds the following new repositories:
+*   P4RT Application
+
+It also has modifications in the following repositories:
+*   SWSS
+*   SONiC Build Image
+
+## SAI API
+There are no changes to SAI. 
+
+For programmable hardware targets, PINS introduces an additional SAI header, saip4ext.h, to map user-defined private P4 extensions to vendor SAI implementations. More details on saip4ext.h can be found in the supplemental document [P4 Extensions for SAI](#broken-todo).
```

Comment:
where can we find saip4ext.h?

Reply:
Here is a pointer: https://github.com/Azure/SONiC/blob/09290206ce8e4a72179614107c63ee396f13c670/doc/pins/p4ext_hld.md#appendix-a

Reply:
Here is the SAI HLD.
https://github.com/opencomputeproject/SAI/pull/1292